gtk/src/toga_gtk/widgets/canvas.py

```diff
@@ -14,6 +14,16 @@ def create(self):
         self.native.interface = self.interface
         self.native.connect("draw", self.gtk_draw_callback)
         self.native.connect("size-allocate", self.gtk_on_size_allocate)
+        self.native.set_events(
+            Gdk.EventMask.BUTTON_PRESS_MASK
+            | Gdk.EventMask.BUTTON_RELEASE_MASK
+            | Gdk.EventMask.POINTER_MOTION_MASK
+        )
+        # count number of active clicks
+        self.clicks = 0
+        # pointer motion event has no button attribute
+        # so track which pointer button was clicked
+        self.button = 0
```

Comment:
It's annoying that these are needed, but there's if there's no other API to get the current mouse state, then I guess this sort of thing is a reasonable approach. 

Cosmetically, these variables should be have names that are a little more specific about exactly why we're tracking clicks and buttons - `button` is an attribute of canvas, but a canvas doesn't have a `button` - it's the most recent mouse event that has a button state. 

However my larger concern is how fragile this might be. What happens if you:

1. Mousedown inside the canvas area
2. Drag outside the canvas area
3. Mouseup outside the canvas area
4. Move back into the canvas area

Will the movements in (4) register as drags?

Reply:
I'm not a GTK expert so I'll dig a little more to make sure there's not a more graceful way to handle this.

Surprisingly, however, the answer to your concern

> Will the movements in (4) register as drags?

is "no." (tested in the canvas example app, at least)

Reply:
Figured it out! There's a `BUTTON_MOTION_MASK` which sends details on whether a pointer button is pressed during motion.